### ISSUE TITLE
DAT_<address> is shown instead of selector when working with dyld shared cache

### ISSUE THREAD

**Issue by lukele** [NationalSecurityAgency/ghidra#7289]
**Describe the bug**
In my instances `DAT_<address>` is shown instead of the selector that is being invoked when working with the dyld_shared_cache.

![Screenshot 2024-12-16 at 15 04 09](https://github.com/user-attachments/assets/1abcd0fa-e006-479f-9bc4-627dcf46a43d)

Using blacktop's `ipsw` however the selector can be read, so it is available:

```
./ipsw dyld a2s /System/Cryptexes/OS/System/Library/dyld/dyld_shared_cache_x86_64  7ffb404adb00            
   • Address location          dylib=/System/Library/Frameworks/MailKit.framework/Versions/A/MailKit section=__DATA_CONST.__objc_selrefs
0x7ffb404adb00: sel_init
```

When right clicking the address and going to References there's also no option to "Add To Program" as was suggested in an article when dealing with the dsc.

**To Reproduce**
1. Go to `File › Open File System`
2. Select the /System/Cryptexes/OS/System/Library/dyld/dyld_shared_cache_x86_64
3. Select MailKit as library to import
4. Choose Format `Mac OS X Mach-O` and in `Options` select `Load Libraries from Disk` to pull in all dependencies
5. Decompile method `-[MEDecodeContext_initWithCoder:]`
6. See `DAT_7ffb404adb00` instead of the selector being called

**Expected behavior**
When the selector / string information is available, that should be shown instead of `DAT_<address>`

**Environment (please complete the following information):**
 - OS: macOS 15.2 (24C101)
 - Java Version: OpenJDK Runtime Environment Temurin-23.0.1+11 (build 23.0.1+11)
 - Ghidra Version: 11.2.1
 - Ghidra Origin: official GitHub distro

**Additional context**
Apologies for not using the correct terminology.

**Comment by nmggithub** [comment]
I'll add to this that functions appear to do the same thing. `FUN_<address>` instead of the external function. This appears to be new to 15.2? I've only noticed it recently.

Apple seems to have dropped some of their shared cache builder code in dyld-1231.3 (which was released with 15.0): https://github.com/apple-oss-distributions/dyld/commit/65bbeed63cec73f313b1d636e63f243964725a9d#diff-c678c688f2fda3e49f76df11d74477d277e708867a662b4335c62a516f91b73d

A later release, dyld-1235.2, seems to touch a "NewSharedCacheBuilder":
https://github.com/apple-oss-distributions/dyld/commit/66c652a1f1f6b7b5266b8bbfd51cb0965d67cc44#diff-3e2b4be1467b97da745caf1f20b8f758acee697b69ca76af41450caa994d4884

Not sure if any of this has to do with these potential regressions, but I thought I'd mention it.

**Comment by nmggithub** [comment]
Looking more into this, this appears to be (in some respects) at least two different issues:

1. All of the trampolines in `__objc_stubs` use pointers to selector strings that are all in one big array in virtual memory. These pointers can be resolved by loading that memory into the `Program`, but I think it might be easier to repoint them back to the strings in `__objc_methname`. They don't start until somewhat in the middle of that section (I'm not sure what the indicators are), but otherwise they selector strings and trampolines appear to be in the same order.
2. External symbols and functions are not resolved to names, due to other optimizations.

Issue 1 is the only one I have looked into deeply, although I would not be surprised if Issue 2 is similar.

**Comment by nmggithub** [comment]
@ryanmkurtz I've been doing a bit of digging into this on my own and I think I found a bug that might be the root cause of this (or at least part of it).

1. The [`SymbolTableCommand`](https://github.com/NationalSecurityAgency/ghidra/blob/Ghidra_11.2.1_build/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/macho/commands/SymbolTableCommand.java) represents a load command in a Mach-O that points to the _file offset_ of the symbol table and the symbol strings.
2. Given that these are file offsets, it wouldn't make sense for them to be negative.
3. However, when playing around with the parsing code, I found that, in some cases, the `symoff` and `stroff` will be negative values.

I'm honestly not sure why that doesn't make it crash, but it does seem to end up with a symbol table that has the right number of symbols (`nsyms` is never big enough to be misinterpreted), but they are all the empty string.

In Kotlin, I can re-parse the symbols like so, and get proper symbol strings:

```kotlin
val (symbolsFileOffset, stringsFileOffset) =
    symbolTableCommand.let {
        Pair(
            it.symbolOffset.toUInt(),
            it.stringTableOffset.toUInt(),
        )
    }
var symbolTable = mutableListOf<NList>()
for (i in 0 until symbolTableCommand.numberOfSymbols) {
    val nList = NList(readerForSymbolTable, machHeader.is32bit)
    nList.initString(readerForSymbolTable, symbolsFileOffset.toLong())
    symbolTable += nList
}
```

I don't think Java has `.toUInt()`, but I think if you change

```java
		symoff = loadCommandReader.readNextInt();
		nsyms = loadCommandReader.readNextInt();
		stroff = loadCommandReader.readNextInt();
		strsize = loadCommandReader.readNextInt();
```

to

```java
		symoff = loadCommandReader.readNextUnsignedInt();
		nsyms = loadCommandReader.readNextInt();
		stroff = loadCommandReader.readNextUnsignedInt();
		strsize = loadCommandReader.readNextInt();
```

it might solve things.

Honestly, they could all be interpreted as unsigned integers, as I don't think any of them would ever be negative.

**Comment by ryanmkurtz** [comment]
Do you guys think it would make sense if it just brought in `libobjc.dylib` by default?  I could have a load option to have it not do that but it seems like you would always want it since there is so much objc stuff everywhere.

**Comment by nmggithub** [comment]
> Do you guys think it would make sense if it just brought in `libobjc.dylib` by default? I could have a load option to have it not do that but it seems like you would always want it since there is so much objc stuff everywhere.

I wouldn't be against this. It's what I'm playing around with in GhidraApple already. Especially for cached dylibs and their `__objc_*list` sections, those often reference `libobjc`. And given that the Objective-C 2 Class analyzer doesn't have a "soft fail" path (as we discussed in #7361), not including `libobjc` would lead to Class analysis failing for a lot of cached dylibs.

**Comment by ryanmkurtz** [comment]
Note that in the original screen shot, the `init` selector is by default too small for Ghirda to decode as a string, so that particular one might not appear correct in the decompiler until a string is made on that data type.

**Comment by lukele** [comment]
Very cool. Thank you very much! I’ll test it and report back!

**Comment by nmggithub** [comment]
@ryanmkurtz Pinging here just so I don't have to make a whole new issue. I have been testing this out in the beta, and while it's good, I find that, since `libobjc` is included by default, its exports (which are many) show up under Exports in the Symbol Tree. I think the markup of exports should be avoided, in general, for additional dylibs mapped into a Program's address space.

**Comment by ryanmkurtz** [comment]
@nmggithub I made this change...can you give it a try?

**Comment by nmggithub** [comment]
@ryanmkurtz Just built it and it looks nice! I don't see the `libobjc` exports anymore.

**Comment by ryanmkurtz** [comment]
Cool, thanks for testing

**Comment by nmggithub** [comment]
> Cool, thanks for testing

Of course!
